jobrunner/run.py

```diff
@@ -24,17 +25,25 @@
 log = logging.getLogger(__name__)
 
 
-def main(exit_when_done=False, raise_on_failure=False):
+def main(exit_when_done=False, raise_on_failure=False, shuffle_jobs=True):
     log.info("jobrunner.run loop started")
     while True:
-        active_jobs = handle_jobs(raise_on_failure=raise_on_failure)
+        active_jobs = handle_jobs(
+            raise_on_failure=raise_on_failure, shuffle_jobs=shuffle_jobs
+        )
         if exit_when_done and len(active_jobs) == 0:
             break
         time.sleep(config.JOB_LOOP_INTERVAL)
 
 
-def handle_jobs(raise_on_failure=False):
+def handle_jobs(raise_on_failure=False, shuffle_jobs=True):
     active_jobs = find_where(Job, state__in=[State.PENDING, State.RUNNING])
+    # Randomising the job order is a crude but effective way to ensure that a
+    # single large job request doesn't hog all the workers. We make this
+    # optional as, when running locally, having jobs run in a predictable order
+    # is preferable
+    if shuffle_jobs:
+        random.shuffle(active_jobs)
```

Comment:
I wonder about shuffling a subset of the jobs, e.g. the top 5 or top len() * 0.5

This preserves a degree of fair queuing - jobs will queue in FIFO order until they enter the top n - whilst keeping the head-of-line unblocking of randomisation.

Randomising the entire list means a job just submitted might gazump one that's been waiting for hours. I also suspect full randomisation might mean highly weighted jobs are often delayed as they lose their chance to be considered before the resources are filled up with small jobs. If there's *some* ordering to the queue, this keeps some fairness next to the unblocking randomness.

Reply:
I'm not sure that will do what we need. The specific case that prompted this was someone submitting 70 odd jobs which they expected to take all weekend to run and this meaning that no one else could run anything until it was finished. I think having a newly submitted job gazump an old job is fine in the circumstances.

It's a good point that highly weighted jobs won't get scheduled if there are loads of of smaller jobs in the queue. I don't think that's a disaster either though. It's an improvement on the status quo. And the queue spends most of the time empty anyway at the moment, so in practice I don't think it will be a problem.